src/teacher.py

```diff
@@ -38,36 +50,23 @@ def _answer_GroundAtomsHold_query(
             holds[ground_atom] = pred.holds(state, ground_atom.objects)
         return GroundAtomsHoldResponse(query, holds)
 
-
-@dataclass(frozen=True, eq=False, repr=False)
-class Query(abc.ABC):
-    """Base class for a Query.
-
-    Has no API.
-    """
-
-
-@dataclass(frozen=True, eq=False, repr=False)
-class Response(abc.ABC):
-    """Base class for a Response to a query.
-
-    All responses contain the Query object itself, for convenience.
-    """
-    query: Query
-
-
-################################################################################
-#                        Query and Response subclasses                         #
-################################################################################
-
-
-@dataclass(frozen=True, eq=False, repr=False)
-class GroundAtomsHoldQuery(Query):
-    """A query for whether ground atoms hold in the state."""
-    ground_atoms: Collection[GroundAtom]
-
-
-@dataclass(frozen=True, eq=False, repr=False)
-class GroundAtomsHoldResponse(Response):
-    """A response to a GroundAtomsHoldQuery, providing boolean answers."""
-    holds: Dict[GroundAtom, bool]
+    def _answer_Demonstration_query(
+            self, state: State,
+            query: DemonstrationQuery) -> DemonstrationResponse:
+        task = Task(state, query.goal)
+        try:
+            policy = self._oracle_approach.solve(task, CFG.timeout)
+        except (ApproachTimeout, ApproachFailure):
+            return DemonstrationResponse(query, None)
+
+        def goal_reached(s: State) -> bool:
+            for ground_atom in query.goal:
```

Comment:
you can use `ground_atom.holds` instead, and do this in one line with an `all(...)`

Reply:
alternatively, you could just use `run_policy_on_task` in utils?